graders/java/autograder.sh

```diff
@@ -42,20 +43,36 @@ fi
 
 RESULTS_TEMP_DIR=$(mktemp -d -p /grade/results)
 RESULTS_TEMP_FILE="$RESULTS_TEMP_DIR/$RANDOM.json"
+SIGNATURE=$(head -c 32 /dev/random | base64)
+
+jq -n --arg results_file "$RESULTS_TEMP_FILE" \
+   --arg compile_output "$STUDENT_COMPILE_OUT" \
+   --arg test_files "$TEST_FILES" \
+   --arg signature "$SIGNATURE" \
+   '{results_file: $results_file, compile_output: $compile_output, test_files: $test_files, signature: $signature}' > /grade/params/params.json
 
 chmod 700 /javagrader
 chmod 711 /grade
 chmod 700 /grade/*
 chmod 711 /grade/results
 chmod -R a+rX /grade/classpath
 chmod 777 $RESULTS_TEMP_DIR
+chmod 777 /grade/params
+chmod 777 /grade/params/params.json
 
 su - sbuser <<EOF
-java -cp "$CLASSPATH" JUnitAutograder "$RESULTS_TEMP_FILE" "$TEST_FILES" "$STUDENT_COMPILE_OUT"
+java -cp "$CLASSPATH" JUnitAutograder
 EOF
 
 if [ -f $RESULTS_TEMP_FILE ] ; then
-    mv $RESULTS_TEMP_FILE $RESULTS_FILE
+    RESULT_SIGNATURE=$(jq -r '.signature' $RESULTS_TEMP_FILE)
+    if [ "$RESULT_SIGNATURE" != "$SIGNATURE" ] ; then
+        echo Expected signature: $SIGNATURE
+        echo Provided signature: $RESULT_SIGNATURE
+        exception "Results did not contain correct signature. Exploit suspected. Please contact the instructor."
```

Comment:
😆 Should we deliberately make this more vague to encourage actual reporting? "Unable to process submission. Please contact the instructor."

Reply:
Along those lines, I'd recommend making the error message completely generic (like 0% grade, compilation error, or server error) with an explanation only appearing on the instructor log side.

Reply:
By generic I just mean indistinguishable from an error that would be emitted by some other code path. Indicating on the frontend whether the signature was the issue or not could be used for fuzzing or something, I imagine.

Reply:
Considering this is open source, I don't think it's that useful to make it too obscure, at the risk of instructors unable to understand why a student would get this error. I'm merging as is, future PRs can address that if necessary.